File: src/test/java/cz/vhromada/catalog/service/PictureServiceTest.java
package cz.vhromada.catalog.service;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import cz.vhromada.catalog.domain.Picture;
import cz.vhromada.catalog.repository.PictureRepository;
import cz.vhromada.catalog.utils.PictureUtils;
import cz.vhromada.common.repository.MovableRepository;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.service.MovableServiceTest;

import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.cache.Cache;

/**
 * A class represents test for class {@link PictureService}.
 *
 * @author Vladimir Hromada
 */
class PictureServiceTest extends MovableServiceTest<Picture> {

    /**
     * Instance of {@link PictureRepository}
     */
    @Mock
    private PictureRepository repository;

    /**
     * Test method for {@link PictureService#PictureService(PictureRepository, Cache)} with null repository for pictures.
     */
    @Test
    void constructor_NullPictureRepository() {
        assertThatThrownBy(() -> new PictureService(null, getCache())).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link PictureService#PictureService(PictureRepository, Cache)} with null cache.
     */
    @Test
    void constructor_NullCache() {
        assertThatThrownBy(() -> new PictureService(repository, null)).isInstanceOf(IllegalArgumentException.class);
    }

    @Override
    protected MovableRepository<Picture> getRepository() {
        return repository;
    }

    @Override
    protected MovableService<Picture> getService() {
        return new PictureService(repository, getCache());
    }

    @Override
    protected String getCacheKey() {
        return "pictures";
    }

    @Override
    protected Picture getItem1() {
        return PictureUtils.newPictureDomain(1);
    }

    @Override
    protected Picture getItem2() {
        return PictureUtils.newPictureDomain(2);
    }

    @Override
    protected Picture getAddItem() {
        return PictureUtils.newPictureDomain(null);
    }

    @Override
    protected Picture getCopyItem() {
        final Picture picture = PictureUtils.newPictureDomain(null);
        picture.setPosition(0);

        return picture;
    }

    @Override
    protected Class<Picture> getItemClass() {
        return Picture.class;
    }

    @Override
    protected void assertDataDeepEquals(final Picture expected, final Picture actual) {
        PictureUtils.assertPictureDeepEquals(expected, actual);
    }

}

File: src/test/kotlin/cz/vhromada/catalog/service/PictureServiceTest.kt
package cz.vhromada.catalog.service

import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
import com.nhaarman.mockitokotlin2.whenever
import cz.vhromada.catalog.domain.Picture
import cz.vhromada.catalog.repository.PictureRepository
import cz.vhromada.catalog.utils.PictureUtils
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.test.service.MovableServiceTest
import cz.vhromada.common.test.utils.TestConstants
import org.mockito.Mock
import org.springframework.data.jpa.repository.JpaRepository

/**
 * A class represents test for class [PictureService].
 *
 * @author Vladimir Hromada
 */
class PictureServiceTest : MovableServiceTest<Picture>() {

    /**
     * Instance of [PictureRepository]
     */
    @Mock
    private lateinit var repository: PictureRepository

    /**
     * Instance of [AccountProvider]
     */
    @Mock
    private lateinit var accountProvider: AccountProvider

    /**
     * Instance of [TimeProvider]
     */
    @Mock
    private lateinit var timeProvider: TimeProvider

    override fun getRepository(): JpaRepository<Picture, Int> {
        return repository
    }

    override fun getAccountProvider(): AccountProvider {
        return accountProvider
    }

    override fun getTimeProvider(): TimeProvider {
        return timeProvider
    }

    override fun getService(): MovableService<Picture> {
        return PictureService(repository, accountProvider, timeProvider, cache)
    }

    override fun getCacheKey(): String {
        return "pictures${TestConstants.ACCOUNT.id}"
    }

    override fun getItem1(): Picture {
        return PictureUtils.newPictureDomain(1)
    }

    override fun getItem2(): Picture {
        return PictureUtils.newPictureDomain(2)
    }

    override fun getAddItem(): Picture {
        return PictureUtils.newPictureDomain(null)
    }

    override fun getCopyItem(): Picture {
        return PictureUtils.newPictureDomain(null)
                .copy(position = 0)
    }

    override fun initAllDataMock(data: List<Picture>) {
        whenever(repository.findByAuditCreatedUser(any())).thenReturn(data)
    }

    override fun verifyAllDataMock() {
        verify(repository).findByAuditCreatedUser(TestConstants.ACCOUNT_ID)
        verifyNoMoreInteractions(repository)
    }

    override fun anyItem(): Picture {
        return any()
    }

    override fun argumentCaptorItem(): KArgumentCaptor<Picture> {
        return argumentCaptor()
    }

    override fun assertDataDeepEquals(expected: Picture, actual: Picture) {
        PictureUtils.assertPictureDeepEquals(expected, actual)
    }

}
